src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -166,27 +178,57 @@ end = struct
           @@ Logger.error t.logger
                !"Could not get the proof of root from the network: %s"
                (Error.to_string_hum e)
-      | Ok peer_root_with_proof -> (
+      | Ok (peer_root_with_proof, scan_state, staged_ledger_merkle_root) -> (
           match%map
             Root_prover.verify ~logger:t.logger ~observed_state:candidate_state
               ~peer_root:peer_root_with_proof
           with
           | Ok (peer_root, peer_best_tip) ->
               t.best_seen_transition <- peer_best_tip ;
               t.current_root <- peer_root ;
-              let ledger_hash =
+              let blockchain_state =
                 let open External_transition in
                 t.current_root |> With_hash.data
                 |> Proof_verified.protocol_state
                 |> Protocol_state.blockchain_state
-                |> Protocol_state.Blockchain_state.snarked_ledger_hash
-                |> Frozen_ledger_hash.to_ledger_hash
               in
-              Root_sync_ledger.new_goal root_sync_ledger ledger_hash
-              |> Fn.const `Syncing
+              let expected_staged_ledger_hash =
+                blockchain_state
+                |> External_transition.Protocol_state.Blockchain_state
+                   .staged_ledger_hash
+              in
+              let received_staged_ledger_hash =
+                Staged_ledger_hash.of_aux_and_ledger_hash
+                  (Staged_ledger_hash.Aux_hash.of_bytes
+                     (Staged_ledger_aux_hash.to_bytes
+                        (Staged_ledger.Scan_state.hash scan_state)))
+                  staged_ledger_merkle_root
+              in
+              if
+                Staged_ledger_hash.equal expected_staged_ledger_hash
+                  received_staged_ledger_hash
+              then
+                let snarked_ledger_hash =
+                  blockchain_state
+                  |> External_transition.Protocol_state.Blockchain_state
+                     .snarked_ledger_hash
+                in
+                Root_sync_ledger.new_goal root_sync_ledger
+                  (Frozen_ledger_hash.to_ledger_hash snarked_ledger_hash)
+                |> Fn.const
+                   @@ `Syncing
+                        { snarked_ledger_hash
+                        ; staged_ledger_merkle_root
+                        ; scan_state }
+              else
+                (* TODO: punish *)
+                Fn.const `Ignored
```

Comment:
This is an anti pattern. If you write code in this style instead of like this example:

```ocaml
else (
  Logger.faulty_peer t.logger "...";
  `Ignored)
```

you can cause subtle bugs since `;` will throw errors (warnings actually, but we mark them as errors) if the type of the left hand side of the `;` is not `unit`. In this style, for instance, imagine you had written the code:

```ocaml
Fn.const `Ignored @@ Logger.faulty_peer t.logger "%d"
```

This would compile fine even though you only partially applied the arguments to `Logger.faulty_peer`, so your program will run without actually logging that line.